components/logins/src/error.rs

```diff
@@ -3,19 +3,49 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use std::ffi::OsString;
+pub type Result<T> = std::result::Result<T, LoginsError>;
+// Functions which are part of the public API should use this Result.
+pub type ApiResult<T> = std::result::Result<T, LoginsStorageError>;
```

Comment:
I'm not sure how that would work out in practice - IIUC, that would mean you'd need to convert from a sqlite error to this internal error at the place where the sqlite error might happen (ie, whereever you'd want to write `?` on a function returning a sqlite error). This would then make it tricky to handle specific errors in the crate in a sane way

eg, consider:

```
// Assume `Result` has our thin "public" error enum, so no explicit serde or sqlite errors
fn outer() -> Result<()> {
  middle()?;
}

fn middle() -> Result<()> {
  inner()?;
}

fn inner() -> Result<()> {
 do_sql()?;
 do_serde()?;
}
```

We need an `Into` our Err for the sql error and another for the serde error, and that conversion would happen at `inner()`. But what if `middle` actually *did* want to handle some sql or serde errors thrown by the inner functions? It now only has the error after transformation into the "Internal" representatation.

(Or maybe I minunderstand your suggestion or how it might be implemented?)

Reply:
Yeah looks like the lost context is somewhat a killer to what I was suggesting :( 

> There does seem to be a fundamental split at the API boundary. The consumer code and the code in the crate care about very different things when it comes to errors.

+1! I can see that and I concede here, I think it'll be worth it to have both (or at the very least worth seeing how Logins shapes up) - it would also be nice to have a small document going over the reasoning for record-keeping!